File: sdk/typescript/src/clients/access.ts
/**
 * Access control resource client.
 *
 * Design Reference: DR-5
 * Requirements: 8.1, 8.2, 8.3
 */

import type { HTTPTransport } from '../transport.js';
import type { AccessResponse, Collaborator } from '../types/index.js';

/**
 * Parse ISO date string to Date object.
 */
function parseDate(dateStr: string): Date {
  return new Date(dateStr.replace(/Z$/, '+00:00'));
}

/**
 * Client for repository access control operations.
 *
 * Design Reference: DR-5
 */
export class AccessClient {
  private transport: HTTPTransport;

  /**
   * Initialize the access client.
   *
   * @param transport - HTTP transport for making requests
   */
  constructor(transport: HTTPTransport) {
    this.transport = transport;
  }

  /**
   * Grant repository access to an agent.
   *
   * Requires admin access to the repository.
   *
   * @param repoId - The repository identifier
   * @param agentId - The agent to grant access to
   * @param role - Access role ("read", "write", or "admin")
   * @returns AccessResponse with action "granted"
   * @throws AuthenticationError if signature is invalid
   * @throws AuthorizationError if not authorized (requires admin)
   * @throws NotFoundError if repository or agent not found
   *
   * Requirements: 8.1
   */
  async grant(
    repoId: string,
    agentId: string,
    role: 'read' | 'write' | 'admin'
  ): Promise<AccessResponse> {
    const response = await this.transport.signedRequest(
      'POST',
      `/v1/repos/${repoId}/access`,
      'access_grant',
      {
        repoId,
        targetAgentId: agentId,
        role,
      }
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return {
      repoId: data.repoId as string,
      agentId: data.agentId as string,
      role: (data.role as string | null) ?? null,
      action: data.action as 'granted' | 'revoked',
    };
  }

  /**
   * Revoke repository access from an agent.
   *
   * Requires admin access to the repository.
   *
   * @param repoId - The repository identifier
   * @param agentId - The agent to revoke access from
   * @returns AccessResponse with action "revoked"
   * @throws AuthenticationError if signature is invalid
   * @throws AuthorizationError if not authorized (requires admin)
   * @throws NotFoundError if repository or agent not found
   *
   * Requirements: 8.2
   */
  async revoke(repoId: string, agentId: string): Promise<AccessResponse> {
    const response = await this.transport.signedRequest(
      'DELETE',
      `/v1/repos/${repoId}/access/${agentId}`,
      'access_revoke',
      {
        repoId,
        targetAgentId: agentId,
      }
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return {
      repoId: data.repoId as string,
      agentId: data.agentId as string,
      role: (data.role as string | null) ?? null,
      action: data.action as 'granted' | 'revoked',
    };
  }

  /**
   * List repository collaborators.
   *
   * @param repoId - The repository identifier
   * @returns List of Collaborator objects with agentId, agentName, role, grantedAt
   * @throws AuthenticationError if signature is invalid
   * @throws NotFoundError if repository not found
   *
   * Requirements: 8.3
   */
  async list(repoId: string): Promise<Collaborator[]> {
    const response = await this.transport.signedRequest(
      'GET',
      `/v1/repos/${repoId}/access`,
      'access_list',
      { repoId }
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    const collaborators = (data.collaborators ?? []) as Record<string, unknown>[];
    return collaborators.map((collab) => ({
      agentId: collab.agentId as string,
      agentName: collab.agentName as string,
      role: collab.role as 'read' | 'write' | 'admin',
      grantedAt: parseDate(collab.grantedAt as string),
    }));
  }
}

File: sdk/rust/src/clients/access.rs
//! Access control resource client.
//!
//! Design Reference: DR-5
//! Requirements: 8.1, 8.2, 8.3

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;

use crate::error::Error;
use crate::transport::HttpTransport;
use crate::types::{AccessResponse, Collaborator};

/// Client for repository access control operations.
pub struct AccessClient {
    transport: Arc<HttpTransport>,
}

impl AccessClient {
    /// Create a new access client.
    pub fn new(transport: Arc<HttpTransport>) -> Self {
        Self { transport }
    }

    /// Grant repository access to an agent.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `agent_id` - The agent to grant access to
    /// * `role` - Role to grant: "read", "write", or "admin"
    ///
    /// # Returns
    ///
    /// AccessResponse with action "granted"
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    ///
    /// Requirements: 8.1
    pub async fn grant(
        &self,
        repo_id: &str,
        agent_id: &str,
        role: &str,
    ) -> Result<AccessResponse, Error> {
        let mut body: HashMap<String, Value> = HashMap::new();
        body.insert("repoId".to_string(), Value::String(repo_id.to_string()));
        body.insert("agentId".to_string(), Value::String(agent_id.to_string()));
        body.insert("role".to_string(), Value::String(role.to_string()));

        let response: Value = self
            .transport
            .signed_request(
                "POST",
                &format!("/v1/repos/{repo_id}/collaborators"),
                "access_grant",
                body,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// Revoke repository access from an agent.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `agent_id` - The agent to revoke access from
    ///
    /// # Returns
    ///
    /// AccessResponse with action "revoked"
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    ///
    /// Requirements: 8.2
    pub async fn revoke(&self, repo_id: &str, agent_id: &str) -> Result<AccessResponse, Error> {
        let mut body: HashMap<String, Value> = HashMap::new();
        body.insert("repoId".to_string(), Value::String(repo_id.to_string()));
        body.insert("agentId".to_string(), Value::String(agent_id.to_string()));

        let response: Value = self
            .transport
            .signed_request(
                "DELETE",
                &format!("/v1/repos/{repo_id}/collaborators/{agent_id}"),
                "access_revoke",
                body,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// List repository collaborators.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    ///
    /// # Returns
    ///
    /// List of Collaborator objects
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    ///
    /// Requirements: 8.3
    pub async fn list(&self, repo_id: &str) -> Result<Vec<Collaborator>, Error> {
        let response: Value = self
            .transport
            .unsigned_request::<Value>(
                "GET",
                &format!("/v1/repos/{repo_id}/collaborators"),
                None,
                None::<&()>,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        let collaborators = data
            .get("collaborators")
            .ok_or_else(|| Error::Http("Missing collaborators in response".to_string()))?;

        serde_json::from_value(collaborators.clone()).map_err(Error::from)
    }
}
